File: integrations/crewai/cortadel_crewai/tools.py
"""CrewAI-native tools that let an agent call Cortadel itself.

These are ordinary ``crewai.tools.BaseTool`` subclasses with pydantic
``args_schema`` models, so they drop into ``Agent(tools=[...])`` alongside any
other CrewAI tool. Use them when you want an agent to *decide* when to search
or store — as opposed to :class:`~cortadel_crewai.CortadelMemory`, which wires
Cortadel in automatically for every turn.

The two are independent: you can use either, or both together.
"""

from __future__ import annotations

from typing import Any, Callable, Optional

from cortadel import AddOptions, SearchOptions
from crewai.tools import BaseTool
from pydantic import BaseModel, Field

from cortadel_crewai._compat import (
    DEFAULT_APP_NAME,
    build_client,
    clamp_top_k,
    guard,
)


class SearchMemorySchema(BaseModel):
    """Arguments for :class:`CortadelSearchMemoryTool`."""

    query: str = Field(
        ...,
        description="Natural-language description of what you are trying to remember.",
    )
    top_k: int = Field(
        default=10,
        description="Maximum number of memories to return (1-50).",
    )


class AddMemoriesSchema(BaseModel):
    """Arguments for :class:`CortadelAddMemoriesTool`."""

    messages: list[str] = Field(
        ...,
        description=(
            "One or more durable facts, decisions, or preferences worth "
            "remembering for future tasks. Keep each one self-contained."
        ),
    )


class _CortadelToolBase(BaseTool):
    """Shared Cortadel client plumbing for the tools below."""

    base_url: Optional[str] = Field(default=None, exclude=True)
    user_id: Optional[str] = Field(default=None, exclude=True)
    # repr=False as well as exclude=True: this is a bearer token, and `exclude`
    # only keeps it out of model_dump(). CrewAI reprs tools when verbose.
    api_key: Optional[str] = Field(default=None, exclude=True, repr=False)
    app_name: str = Field(default=DEFAULT_APP_NAME, exclude=True)
    client: Any = Field(default=None, exclude=True)
    raise_on_error: bool = Field(default=False, exclude=True)
    on_error: Optional[Callable[[BaseException], None]] = Field(
        default=None, exclude=True, repr=False
    )

    def model_post_init(self, __context: Any) -> None:
        super().model_post_init(__context)
        if self.client is None:
            self.client = build_client(
                base_url=self.base_url,
                user_id=self.user_id,
                api_key=self.api_key,
                app_name=self.app_name,
            )


class CortadelSearchMemoryTool(_CortadelToolBase):
    """Let an agent search its long-term Cortadel memory."""

    name: str = "search_memory"
    description: str = (
        "Search long-term memory for facts, preferences, and decisions recorded "
        "in earlier sessions. Use this before answering anything that depends on "
        "prior context. Input is a natural-language query."
    )
    args_schema: type[BaseModel] = SearchMemorySchema

    mode: str = Field(default="hybrid", exclude=True)
    rerank: Optional[str] = Field(default=None, exclude=True)

    def _run(self, query: str, top_k: int = 10, **kwargs: Any) -> str:
        options = SearchOptions(
            top_k=clamp_top_k(top_k), mode=self.mode, rerank=self.rerank
        )
        results = guard(
            "search",
            lambda: self.client.search(query, options),
            None,
            raise_on_error=self.raise_on_error,
            on_error=self.on_error,
        )
        if results is None:
            return "Memory is temporarily unavailable; continue without it."
        if not results.results:
            return "No relevant memories found."

        lines = []
        for hit in results.results:
            prefix = "- "
            if hit.categories:
                prefix = f"- [{', '.join(hit.categories)}] "
            lines.append(f"{prefix}{hit.content}")
        return "Relevant memories:\n" + "\n".join(lines)


class CortadelAddMemoriesTool(_CortadelToolBase):
    """Let an agent write durable facts into Cortadel."""

    name: str = "add_memories"
    description: str = (
        "Save durable facts, preferences, or decisions to long-term memory so "
        "they are available in future sessions. Pass one or more short, "
        "self-contained statements."
    )
    args_schema: type[BaseModel] = AddMemoriesSchema

    infer: bool = Field(default=True, exclude=True)
    memory_type: Optional[str] = Field(default=None, exclude=True)

    def _run(self, messages: list[str] | str, **kwargs: Any) -> str:
        if isinstance(messages, str):
            messages = [messages]
        items = [m for m in messages if m and m.strip()]
        if not items:
            return "Nothing to save."

        options = AddOptions(
            app=self.app_name, infer=self.infer, memory_type=self.memory_type
        )
        saved = 0
        duplicates = 0
        for text in items:
            created = guard(
                "add",
                lambda text=text: self.client.add(text, options),
                None,
                raise_on_error=self.raise_on_error,
                on_error=self.on_error,
            )
            if created is None:
                continue
            # A 2xx does not mean a new memory was written — check the event.
            if created.event == "SKIP_DUPLICATE":
                duplicates += 1
            else:
                saved += 1

        if saved == 0 and duplicates == 0:
            return "Memory is temporarily unavailable; nothing was saved."
        parts = [f"Saved {saved} memory item(s)."]
        if duplicates:
            parts.append(f"{duplicates} were already known and were skipped.")
        return " ".join(parts)


def cortadel_tools(
    base_url: Optional[str] = None,
    user_id: Optional[str] = None,
    api_key: Optional[str] = None,
    app_name: str = DEFAULT_APP_NAME,
    client: Any = None,
    raise_on_error: bool = False,
    on_error: Optional[Callable[[BaseException], None]] = None,
) -> list[BaseTool]:
    """Build both Cortadel tools sharing one client.

    Mirrors CrewAI's own ``create_memory_tools`` helper::

        agent = Agent(role=..., goal=..., backstory=...,
                      tools=cortadel_tools(user_id="e2e-crewai-demo"))

    Args:
        raise_on_error: When True, Cortadel failures propagate to the agent loop
            instead of degrading to a "memory is unavailable" tool result.
        on_error: Called with the exception on every Cortadel failure. Setting it
            replaces the default warning log.
    """
    shared = client if client is not None else build_client(
        base_url=base_url, user_id=user_id, api_key=api_key, app_name=app_name
    )
    return [
        CortadelSearchMemoryTool(
            client=shared,
            app_name=app_name,
            raise_on_error=raise_on_error,
            on_error=on_error,
        ),
        CortadelAddMemoriesTool(
            client=shared,
            app_name=app_name,
            raise_on_error=raise_on_error,
            on_error=on_error,
        ),
    ]

File: integrations/vercel-ai-sdk/src/tools.ts
import { tool } from "ai";
import type { Tool, ToolSet } from "ai";
import { z } from "zod";

import { ClientResolver } from "./connection.js";
import type {
  CortadelConnectionOptions,
  CortadelErrorContext,
  CortadelMemoryClient,
} from "./types.js";

/** Matches the Cortadel SDK's own `SearchOptions.topK` default. */
const DEFAULT_TOOL_TOP_K = 10;

/** Search knobs for the `search_memory` tool. */
export interface CortadelSearchToolOptions {
  /**
   * Default result count when the model does not ask for one (1–50). Defaults to `10`, the
   * Cortadel SDK's own `SearchOptions` default — a tool the model called on purpose can afford a
   * wider net than the middleware's automatic injection, which pays for every hit in every prompt.
   */
  topK?: number;
  /** Search mode. Defaults to `"hybrid"`. */
  mode?: "hybrid" | "text" | "vector";
  /** Set to `"cross_encoder"` to rerank with the server's local cross-encoder. */
  rerank?: "cross_encoder";
  /** Restrict results to one cognitive type. */
  memoryType?: "episodic" | "semantic" | "procedural";
  /** Restrict results to one session. */
  sessionId?: string;
  /** Drop hits whose `rrfScore` is below this. Unscored hits are kept. */
  minScore?: number;
}

/** Write knobs for the `add_memories` tool. */
export interface CortadelAddToolOptions {
  /**
   * App name recorded as the creator of each memory (the SDK's `AddOptions.app`). Unset by
   * default, which leaves the attribution to the server.
   *
   * Distinct from the connection-level `appName`, which the SDK records for access logging on
   * *searches* only and never applies to a write.
   */
  app?: string;
  /**
   * When `false`, store text verbatim and skip background entity/category extraction. Server-side
   * dedup still applies. Defaults to the server default (`true`).
   */
  infer?: boolean;
  /** Pin every stored memory to a cognitive type. */
  memoryType?: "episodic" | "semantic" | "procedural";
  /** Metadata attached to every stored memory. */
  metadata?: Record<string, unknown>;
}

/** Options for {@link cortadelTools}. */
export interface CortadelToolsOptions extends CortadelConnectionOptions {
  search?: CortadelSearchToolOptions;
  add?: CortadelAddToolOptions;

  /**
   * Observe a failed tool call. Called whether or not
   * {@link CortadelToolsOptions.throwOnError} is set — it reports, it does not decide control flow.
   *
   * With no callback and no `throwOnError`, a swallowed failure is logged through `console.warn`
   * rather than vanishing silently.
   */
  onError?: (error: unknown, context: CortadelErrorContext) => void;

  /**
   * Throw out of the tool instead of returning a result carrying `error`. Defaults to `false`: a
   * memory outage should leave the model able to answer without memory, not abort the run.
   *
   * With `true` the AI SDK sees the rejection and surfaces it as a tool error, which is what you
   * want when an answer built without memory would be worse than no answer.
   */
  throwOnError?: boolean;
}

/** One hit returned to the model by `search_memory`. */
export interface RecalledMemory {
  id: string;
  content: string;
  /** Fused relevance score (RRF), when the server returned one. */
  score?: number;
  /** ISO 8601 creation timestamp, when the server returned one. */
  createdAt?: string;
}

/** Result of `search_memory`. Failures arrive as `error` rather than a throw, unless `throwOnError`. */
export interface SearchMemoryResult {
  memories: RecalledMemory[];
  count: number;
  error?: string;
}

/** Result of `add_memories`. Failures arrive as `error` rather than a throw, unless `throwOnError`. */
export interface AddMemoriesResult {
  /** Memories the server actually wrote. */
  stored: number;
  /** Memories the server folded into an existing one (`SKIP_DUPLICATE`). */
  duplicates: number;
  /** Ids of everything the server acknowledged. */
  ids: string[];
  error?: string;
}

const searchMemoryInput = z.object({
  query: z
    .string()
    .min(1)
    .describe(
      "What to recall, in natural language. Write it as the question you want answered, not as " +
        "keywords — search is hybrid (BM25 + vector), so full phrasing retrieves better.",
    ),
  topK: z
    .number()
    .int()
    .min(1)
    .max(50)
    .optional()
    .describe("How many memories to return. Leave unset for the configured default."),
});

const addMemoriesInput = z.object({
  memories: z
    .array(z.string().min(1))
    .min(1)
    .describe(
      "Durable facts worth remembering for later conversations, one self-contained sentence " +
        "each. Write them so they still make sense with no surrounding context. Do not store " +
        "transient details of the current task.",
    ),
});

/** The tools this package registers, keyed by the names the model sees. */
export interface CortadelToolSet extends ToolSet {
  search_memory: Tool<z.infer<typeof searchMemoryInput>, SearchMemoryResult>;
  add_memories: Tool<z.infer<typeof addMemoriesInput>, AddMemoriesResult>;
}

function messageOf(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }
  return String(error);
}

/**
 * Cortadel memory as AI SDK tools the model can call for itself.
 *
 * Returns a `ToolSet` with `search_memory` and `add_memories`, named to match Cortadel's MCP tools
 * so an agent behaves the same whichever surface it reaches memory through. Spread it into
 * `generateText` / `streamText` / `Agent`:
 *
 * ```ts
 * const result = await generateText({
 *   model: openai("gpt-5"),
 *   tools: { ...cortadelTools({ baseUrl: "http://localhost:3001", userId: "e2e-alice" }) },
 *   prompt: "What do you remember about my deployment setup?",
 * });
 * ```
 *
 * The tool set is bound to one user id, because a Cortadel client is: there is no per-call user
 * parameter, and deliberately so — the model can never be talked into reading another user's
 * memories, since the user id is not part of any tool's input schema. For a multi-tenant server,
 * build a tool set per request; it is a plain object literal and costs nothing.
 */
export function cortadelTools(options: CortadelToolsOptions): CortadelToolSet {
  const resolver = new ClientResolver(options);
  // No per-request override exists here — a tool set is bound to one user id — so the resolver is
  // asked for the configured one. Its constructor has already rejected a config that has neither a
  // `userId` nor a client to take one from, which is what the cast records.
  const userId = resolver.resolveUserId() as string;
  const client: CortadelMemoryClient = resolver.resolve(userId);

  const searchOptions = options.search ?? {};
  const addOptions = options.add ?? {};
  const onError = options.onError;
  const throwOnError = options.throwOnError ?? false;

  /**
   * Reports a failed tool call, then decides its fate: rethrow when the caller asked for that,
   * otherwise `console.warn` if nothing else is watching, so a silent memory outage is impossible.
   */
  function report(error: unknown, context: CortadelErrorContext): void {
    if (onError != null) {
      try {
        onError(error, context);
      } catch {
        // A broken error handler must not be more fatal than the error it was handed.
      }
    }
    if (throwOnError) {
      throw error;
    }
    if (onError == null) {
      console.warn(
        `Cortadel: ${context.phase} failed for user "${context.userId}" — the tool returned an error to the model.`,
        error,
      );
    }
  }

  return {
    search_memory: tool({
      description:
        "Search this user's long-term memory for facts recorded in earlier conversations. Call " +
        "it before answering anything that depends on what you were told previously — " +
        "preferences, decisions, ongoing work, prior context. Returns an empty list when nothing " +
        "relevant is stored, which means you have no memory of it, not that it is false.",
      inputSchema: searchMemoryInput,
      async execute({ query, topK }, { abortSignal }): Promise<SearchMemoryResult> {
        try {
          const found = await client.search(
            query,
            {
              topK: topK ?? searchOptions.topK ?? DEFAULT_TOOL_TOP_K,
              mode: searchOptions.mode ?? "hybrid",
              rerank: searchOptions.rerank,
              memoryType: searchOptions.memoryType,
              sessionId: searchOptions.sessionId,
            },
            abortSignal,
          );

          const memories: RecalledMemory[] = (found.results ?? [])
            .filter(
              (hit) =>
                searchOptions.minScore == null ||
                hit.rrfScore == null ||
                hit.rrfScore >= searchOptions.minScore,
            )
            .map((hit) => ({
              id: hit.id,
              content: hit.content,
              ...(hit.rrfScore == null ? {} : { score: hit.rrfScore }),
              ...(hit.createdAt == null ? {} : { createdAt: hit.createdAt }),
            }));

          return { memories, count: memories.length };
        } catch (error) {
          report(error, { phase: "recall", userId });
          return { memories: [], count: 0, error: messageOf(error) };
        }
      },
    }),

    add_memories: tool({
      description:
        "Store durable facts in this user's long-term memory so they survive into future " +
        "conversations. Use it when the user states a preference, a decision, or a fact about " +
        "themselves or their work that will still matter later. The server deduplicates against " +
        "what is already stored, so a repeat is harmless.",
      inputSchema: addMemoriesInput,
      async execute({ memories }, { abortSignal }): Promise<AddMemoriesResult> {
        const ids: string[] = [];
        let stored = 0;
        let duplicates = 0;

        try {
          // Sequential on purpose: the server dedups each write against what is already stored,
          // and firing near-identical facts in parallel races that check.
          for (const text of memories) {
            const created = await client.add(
              text,
              {
                app: addOptions.app,
                infer: addOptions.infer,
                memoryType: addOptions.memoryType,
                metadata: addOptions.metadata,
              },
              abortSignal,
            );

            if (created.id) {
              ids.push(created.id);
            }
            if (created.event === "SKIP_DUPLICATE") {
              duplicates += 1;
            } else {
              stored += 1;
            }
          }

          return { stored, duplicates, ids };
        } catch (error) {
          report(error, { phase: "persist", userId });
          return { stored, duplicates, ids, error: messageOf(error) };
        }
      },
    }),
  };
}
